src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -131,14 +133,67 @@ public boolean equals(Object obj) {
      *
      * @param id the id or {@code null}.
      */
+    @Deprecated
     protected SCMSource(@CheckForNull String id) {
         this.id = id == null ? UUID.randomUUID().toString() : id;
     }
 
     /**
-     * The ID of this source. The ID is not related to anything at all.
+     * Constructor.
+     */
+    public SCMSource() {
+    }
+
+    /**
+     * Sets the ID that is used to ensure this {@link SCMSource} can be retrieved from its {@link SCMSourceOwner}
+     * provided that this {@link SCMSource} does not already {@link #hasId()}.
+     * <p>
+     * Note this is a <strong>one-shot</strong> setter. If {@link #getId()} is called first, then its value will stick,
+     * otherwise the first call to {@link #setId(String)} will stick.
+     *
+     * @param id the ID.
+     * @see #hasId()
+     * @see #getId()
+     * @since 2.2.0
+     */
+    @DataBoundSetter
+    public final synchronized void setId(@CheckForNull String id) {
+        if (this.id == null) {
+            this.id = id;
+        } else {
+            System.out.println("Go away");
+        }
+    }
+
+    /**
+     * Variant of {@link #setId(String)} that can be useful for method chaining.
```

Comment:
Do Stapler & `structs` not support returning values from `@DataBoundSetter`? Should be fixed if not.